### PR TITLE
feature(session): emitToUsers() - allow multiple sockets per session AND add options.origins

### PR DESCRIPTION
**Pull request by andreialecu**
Also allows multiple sessions for an user to work properly with emitToUsers().

Improves https://github.com/deployd/deployd/pull/569 and replaces https://github.com/deployd/deployd/pull/557

### CODE REVIEW

File: lib/session.js [hunk 1]
```diff
@@ -28,43 +28,46 @@ function SessionStore(namespace, db, sockets, options) {
...
-        // index sockets against their session id
-        socketIndex[sid] = socket;
-        if(sessionIndex[sid])
-          sessionIndex[sid].rawSocket = socket ;
-        socketQueue.emit(sid, socket);
+        var session = sessionIndex[sid];
+        if (session) {
+          // index sockets against their session id
+          socketIndex[sid] = socketIndex[sid] || {};
+          socketIndex[sid][client.id] = client;
```
Comment (shaggybb) [inline lib/session.js]: To understand that ¿Is client.id a property of socket?

**Comment by shaggybb** [comment]
First feedback:
I've copied this two files in my project and after passing the tests, I started my project and any change through dpd.collection.on does not trigger an event.

It is very rare, I will check code with node-inspector to understand where is the problem

File: lib/session.js (hunk 1 above)
Comment (andreialecu) [inline lib/session.js]: Yes, it's the socket.io socket id.

**Comment by andreialecu** [comment]
@shaggybb I pushed a new commit with some further changes.

PS: You can use `npm install andreialecu/deployd#refactor-session2` to use my branch/PR and test instead of manually copying files

**Comment by shaggybb** [comment]
@andreialecu In general everything works great excepting in case when the app emits filtering by only one item. 

Tests Round 1:
Test 1 (Chrome: [Tab + Duplicate tab]):
- Open chrome and loggin with user A.
- Duplicate tab (it duplicates session of tab)
Check 
```javascript
var users = [me.id]; // user A
emit(dpd.users, {id:{$in:users}}, "messages:created", this); // FAIL
emit(dpd.users, {id:me.id}, "messages:created", this);  //SUCCESS
```
Test 2 (Chrome: [Tab + Duplicate tab + new tab]):
- Open chrome and loggin with user A.
- Duplicate tab (it duplicates session of tab)
- New tab and set location
Check 
```javascript
var users = [me.id]; // user A
emit(dpd.users, {id:{$in:users}}, "messages:created", this); // FAIL
emit(dpd.users, {id:me.id}, "messages:created", this);  //SUCCESS
```
Test 3 (Firefox + Chrome:[Tab + Duplicate tab + new tab]):
- Open chrome and loggin with user A.
- Duplicate tab (it duplicates session of tab)
- New tab and set location
- Open firefox and loggin with user A.
Check 
```javascript
var users = [me.id]; // user A
emit(dpd.users, {id:{$in:users}}, "messages:created", this); // FAIL
emit(dpd.users, {id:me.id}, "messages:created", this);  //SUCCESS
```

Test 4 (Firefox + Chrome: [Tab + Duplicate tab + new tab] + Chrome in Private mode):
- Open chrome and loggin with user A.
- Duplicate tab (it duplicates session of tab)
- New tab and set location
- Open firefox and loggin with user A.
- Open chrome in private mode and loggin with user B
Check 
```javascript
var users = [me.id];  // user A
emit(dpd.users, {id:{$in:users}}, "messages:created", this); // FAIL
emit(dpd.users, {id:me.id}, "messages:created", this);  //SUCCESS
```

I've checked that when emitToUser get only only user it does not emit per socket.
Solution:
```javascript
  this.emitToUsers = function(collection, query, event, data) {
    collection.get(query, function(users) {      
      var userSession;
      if(users && users.id) {
        users = [users]; // Converts from single object to array of object. This avoids code repetition and allow to iterate sockets with next lines of code.
        // userSession = userSessionIndex[users.id];
        // if(userSession && userSession.socket) {
        //   userSession.socket.emit(event, data);
        // }
        // return;
      }
      users.forEach(function(u) {
        var userSessions = userSessionIndex[u.id];        
        // emit to sessions online
        for (var key in userSessions) {
          userSession = userSessions[key];
          if (userSession && userSession.socket) {
            userSession.socket.emit(event, data);
          }
        }
      });
    });
  };
```

After this change everything works perfectly!
Also I've tested general emit and works perfectly.

Really really good job!

**Comment by andreialecu** [comment]
@shaggybb You are right, I just pushed the fix. Thanks!

**Comment by shaggybb** [comment]
@andreialecu, could you tell me when this PR will be present in the master or in a near version?

Thanks for your time!

**Comment by andreialecu** [comment]
**Important Note: This PR also strictly rejects unknown origins from receiving responses from deployd. Until now all origins were accepted.**

Use something like:

```javascript
var server = deployd({
  port: process.env.PORT || 1337,
  env: 'production',
  origins: [ "http://localhost:3000", "http://myproductionsite.com" ],
  db: { connectionString: process.env.MONGO_URI }
});
```

In your deployd configuration from now on to whitelist your allowed origins.

**If you do not need to use CORS (you use the same origin) you are unaffected by this.**

**Comment by ericfong** [comment]
how about from cordova app or curl?

**Comment by andreialecu** [comment]
@ericfong What do you mean? If you're referring to the origin fix: 

This is simply a CORS option which web browsers respect to prevent cross site request forgery. Curl or other download methods should be unaffected by this.

**Comment by NicolasRitouet** [comment]
Looks good, I'll merge this, anybody against it?

File: lib/session.js [hunk 2]
```diff
@@ -150,8 +158,8 @@ SessionStore.prototype.createSession = function(sid, fn) {
 /**
 * Get the already created session
 */
-SessionStore.prototype.getSession = function (uid) {
-  return userSessionIndex[uid];
+SessionStore.prototype.getSession = function (uid, sid) {
```
Comment (ericfong) [inline lib/session.js]: May be a breaking change for some code.

File: lib/util/http.js [hunk 3]
```diff
@@ -8,12 +8,12 @@ var Cookies = require('cookies')
  * A utility for setting up a request and response.
  */
 
-exports.setup = function(req, res, next) {
+exports.setup = function(options, req, res, next) {
```
Comment (ericfong) [inline lib/util/http.js]: May become a breaking change

File: lib/session.js (hunk 2 above)
Comment (andreialecu) [inline lib/session.js]: Previous implementation was broken, a single uid can have multiple sessions. New signature is more correct.

File: lib/util/http.js (hunk 3 above)
Comment (andreialecu) [inline lib/util/http.js]: http.js is not documented as a public API, no breaking change here I think.

File: lib/session.js (hunk 2 above)
Comment (ericfong) [inline lib/session.js]: Yes. I also support this change. Just want to remark it and let people know.

**Comment by ericfong** [comment]
Thanks for your contribution!
Landed as f67ee1ecc59fd8825059793462fab40fc89d2faf

**Pull request closed without merging**